ssl/statem/statem_dtls.c

```diff
@@ -100,18 +96,65 @@ void dtls1_hm_fragment_free(hm_fragment *frag)
     OPENSSL_free(frag);
 }
 
+static int dtls1_write_hm_header(unsigned char *msgheaderstart,
+                                 unsigned char msg_type, size_t msg_len,
+                                 unsigned short msg_seq, size_t fragoff,
+                                 size_t fraglen)
+{
+    WPACKET msgheader;
+    size_t msgheaderlen;
+
+    if (!WPACKET_init_static_len(&msgheader, msgheaderstart,
+                                 DTLS1_HM_HEADER_LENGTH, 0)
+        || !WPACKET_put_bytes_u8(&msgheader, msg_type)
+        || !WPACKET_put_bytes_u24(&msgheader, msg_len)
+        || !WPACKET_put_bytes_u16(&msgheader, msg_seq)
+        || !WPACKET_put_bytes_u24(&msgheader, fragoff)
+        || !WPACKET_put_bytes_u24(&msgheader, fraglen)
+        || !WPACKET_get_total_written(&msgheader, &msgheaderlen)
+        || msgheaderlen != DTLS1_HM_HEADER_LENGTH
+        || !WPACKET_finish(&msgheader))
+        return 0;
+
+    return 1;
+}
+
 /*
  * send s->init_buf in records of type 'type' (SSL3_RT_HANDSHAKE or
  * SSL3_RT_CHANGE_CIPHER_SPEC)
+ *
+ * When sending a fragmented handshake message this function will re-use
+ * s->init_buf->data but overwrite previously sent data to fill out the handshake
+ * message header for the next fragment.
+ *
+ * E.g.
+ * |-------------------------s->init_buf->data------------------------------|
+ * |-- header1 --||-- fragment1 --|
+ *                  |-- header2 --||-- fragment2 --|
+ *                                   |-- header3 --||-- fragment3 --|
+ *                                                                 .........
  */
 int dtls1_do_write(SSL_CONNECTION *s, uint8_t type)
 {
     int ret;
     size_t written;
     size_t curr_mtu;
     int retry = 1;
-    size_t len, frag_off, overhead, used_len;
+    size_t len, overhead, used_len, msg_len;
     SSL *ssl = SSL_CONNECTION_GET_SSL(s);
+    unsigned char *data = (unsigned char *)s->init_buf->data;
+    unsigned short msg_seq = s->d1->w_msg_hdr.seq;
+    unsigned char msg_type;
+
+    if (type == SSL3_RT_HANDSHAKE) {
+        msg_type = *data++;
+        l3n2(data, msg_len);
+    } else if (ossl_assert(type == SSL3_RT_CHANGE_CIPHER_SPEC)) {
+        msg_type = SSL3_MT_CCS;
+        msg_len = 0; /* SSL3_RT_CHANGE_CIPHER_SPEC */
+    } else
+        /* Other record types are not supported */
+        return -1;
```

Comment:
nit: Code style says that if one branch uses `{}` then they all must - so please use them here.